File: Haskell/4-vestibular.hs
import Text.Printf

calculateResults :: [Int] -> [String] -> IO ()
calculateResults respostas nomes = do
  let pontuacao = map (calcularPontuacao respostas) nomes
      pontuacaoMax = maximum pontuacao
      pontuacaoMin = minimum pontuacao
      halfCorrectCount = length $ filter (\s -> s > 5) pontuacao
      pontuacaoTotal = fromIntegral $ sum pontuacao
      n = length nomes

  let nomesSorteados = sort nomes

  mapM_ (\(name, pontuacao) -> putStrLn $ name ++ " " ++ show (round pontuacaoTotal)) (zip nomesSorteados pontuacao)

  putStrLn $ show pontuacaoMax
  putStrLn $ show pontuacaoMin
  printf "%.1f%%\n" ((fromIntegral halfCorrectCount / fromIntegral n) * 100)

calcularPontuacao :: [Int] -> String -> Int
calcularPontuacao respostas participante = length $ filter (\(a, b) -> a == b) $ zip respostas (map (\x -> read [x] :: Int) (drop 1))

File: Java/4-vestibular.java
import java.util.Arrays;
import java.util.Scanner;

public class vestibular {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int[] respostas = new int[10];
        for (int i = 0; i < 10; i++) {
            respostas[i] = scanner.nextInt();
        }

        int n = 0;
        String[] nomes = new String[100];

        while (true) {
            nomes[n] = scanner.next();
            if (nomes[n].equals("*")) {
                break;
            }
            n++;
        }

        calculateResults(respostas, nomes, n);
    }

    private static void calculateResults(int[] respostas, String[] nomes, int n) {
        int pontuacaoMax = -1, pontuacaoMin = 11, halfCorrectCount = 0;
        double pontuacaoTotal = 0.0;

        for (int i = 0; i < n; i++) {
            int pontuacao = 0;
            for (int j = 0; j < 10; j++) {
                if (respostas[j] == Character.getNumericValue(nomes[i].charAt(j + 1))) {
                    pontuacao++;
                }
            }

            if (pontuacao > pontuacaoMax) {
                pontuacaoMax = pontuacao;
            }
            if (pontuacao < pontuacaoMin) {
                pontuacaoMin = pontuacao;
            }
            if (pontuacao > 5) {
                halfCorrectCount++;
            }

            pontuacaoTotal += pontuacao;
        }

        Arrays.sort(nomes, 0, n);

        for (int i = 0; i < n; i++) {
            System.out.println(nomes[i] + " " + (int) pontuacaoTotal);
        }

        System.out.println(pontuacaoMax);
        System.out.println(pontuacaoMin);
        System.out.printf("%.1f%%\n", ((double) halfCorrectCount / n) * 100);
    }
}
